### PR TITLE
feat: implement turnstile on signup

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?
Implements turnstile captcha to verify signup requests are not coming from bots 

EDIT:

Appearance mode is now https://developers.cloudflare.com/turnstile/get-started/client-side-rendering/#appearance-modes set to interactive-only -> " If appearance is set to interaction-only, the widget will become only visible in cases where an interaction is required"

Fixes #13913

https://github.com/calcom/cal.com/assets/55134778/9b33def6-97b9-4dca-ade7-fe7d148d4b00

## Type of change
- New feature (non-breaking change which adds functionality)

## How should this be tested?

- Are there environment variables that should be set?
> Cloudflare Turnstile
NEXT_PUBLIC_CLOUDFLARE_SITEKEY=
CLOUDFLARE_TURNSTILE_SECRET=
- Any other important info that could help to test that PR
> Please setup turnstile by logging in to a cloudflare account - clicking turnstile on the left and side. Create a site and add `localhost` as the domain without the port. Set the env variables above.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -110,6 +110,8 @@ NEXT_PUBLIC_HELPSCOUT_KEY=
 NEXT_PUBLIC_FRESHCHAT_TOKEN=
 NEXT_PUBLIC_FRESHCHAT_HOST=
 
+NEXT_PUBLIC_CLOUDFLARE_SITEKEY=
+
```
Comment (sean-brydon) [inline .env.example:114]:
```suggestion
```

File: apps/web/components/auth/Turnstile.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import type { TurnstileProps } from "react-turnstile";
+import Turnstile from "react-turnstile";
+
+import { CLOUDFLARE_SITE_ID } from "@calcom/lib/constants";
+
+type Props = Omit<TurnstileProps, "sitekey">;
+
+export default function TurnstileWidget(props: Props) {
+  if (!CLOUDFLARE_SITE_ID || process.env.NEXT_PUBLIC_IS_E2E) return null;
```
Comment (sean-brydon) [inline apps/web/components/auth/Turnstile.tsx:9]: Hide in E2E or if there is no cloudflare site id

File: apps/web/lib/signup/getServerSideProps.tsx [hunk 3]
```diff
@@ -25,13 +25,12 @@ const querySchema = z.object({
...
   const prisma = await import("@calcom/prisma").then((mod) => mod.default);
   const flags = await getFeatureFlagMap(prisma);
-  const ssr = await ssrInit(ctx);
+  await ssrInit(ctx);
   const token = z.string().optional().parse(ctx.query.token);
 
   const props = {
     isGoogleLoginEnabled: IS_GOOGLE_LOGIN_ENABLED,
     isSAMLLoginEnabled,
-    trpcState: ssr.dehydrate(),
```
Comment (sean-brydon) [inline apps/web/lib/signup/getServerSideProps.tsx:34]: This wasnt being used and was causing signup to be too large

File: apps/web/pages/signup.tsx [hunk 4]
```diff
@@ -32,10 +40,15 @@ import { getServerSideProps } from "@lib/signup/getServerSideProps";
 
 import PageWrapper from "@components/PageWrapper";
 
+// import TurnstileCaptcha from "@components/auth/Turnstile";
+
```
Comment (sean-brydon) [inline apps/web/pages/signup.tsx:44]:
```suggestion
```

File: apps/web/pages/signup.tsx [hunk 5]
```diff
@@ -202,7 +215,8 @@ export default function Signup({
 
   const isOrgInviteByLink = orgSlug && !prepopulateFormValues?.username;
 
-  const signUp: SubmitHandler<FormValues> = async (data) => {
+  const signUp: SubmitHandler<FormValues> = async (_data) => {
+    const { cfToken, ...data } = _data;
```
Comment (sean-brydon) [inline apps/web/pages/signup.tsx:219]: Remove the cfToken from the data payload so we can set it on a header

**Review by Udit-takkar — COMMENTED** [review]
Nice. should we also use this on the booking page?

**Comment by sean-brydon** [comment]
> Nice. should we also use this on the booking page?

I was actually just thinking about this 20 mins ago - I don't see any reason why we shouldn't

@calcom/foundation

**Comment by keithwillcode** [comment]
> > Nice. should we also use this on the booking page?
> 
> I was actually just thinking about this 20 mins ago - I don't see any reason why we shouldn't
> 
> @calcom/foundation

Curious what @PeerRich and @ciaranha think but my current thinking is this should only be added to the booking page if non-intrusive and not visible.

**Comment by ciaranha** [comment]
> > > Nice. should we also use this on the booking page?
> 
> > 
> 
> > I was actually just thinking about this 20 mins ago - I don't see any reason why we shouldn't
> 
> > 
> 
> > @calcom/foundation
> 
> 
> 
> Curious what @PeerRich and @ciaranha think but my current thinking is this should only be added to the booking page if non-intrusive and not visible. 

Fully agree on the not visible part. It's not pretty.

File: packages/lib/server/checkCfTurnstileToken.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { HttpError } from "../http-error";
+
+const TUNRSTILE_SECRET_ID = process.env.CLOUDFLARE_TURNSTILE_SECRET;
+
+export async function checkCfTurnstileToken({ token, remoteIp }: { token?: string; remoteIp: string }) {
+  // This means the instant doesnt have turnstile enabled - we skip the check and just return success.
+  // OR the instance is running in CI so we skip these checks also
```
Comment (adasarpan404) [inline packages/lib/server/checkCfTurnstileToken.ts:7]:
```
  if (!remoteIp || !/^(\d{1,3}\.){3}\d{1,3}$/.test(remoteIp)) {
    throw new HttpError({ statusCode: 400, message: "Invalid remote IP address" });
  }
```
This ensures that remoteIp is always provided and is a valid IP address.

File: packages/lib/server/checkCfTurnstileToken.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { HttpError } from "../http-error";
+
+const TUNRSTILE_SECRET_ID = process.env.CLOUDFLARE_TURNSTILE_SECRET;
```
Comment (adasarpan404) [inline packages/lib/server/checkCfTurnstileToken.ts:3]: There's a typo in the variable name `TUNRSTILE_SECRET_ID`, it should be `TURNSTILE_SECRET_ID`

File: packages/lib/server/checkCfTurnstileToken.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    return {
+      success: true,
+    };
+  }
+
+  if (!token) {
+    throw new HttpError({ statusCode: 401, message: "Invalid cloudflare token" });
+  }
+
+  const form = new URLSearchParams();
```
Comment (adasarpan404) [inline packages/lib/server/checkCfTurnstileToken.ts:18]:
```javascript
  try {
    const form = new URLSearchParams();
    form.append("secret", TURNSTILE_SECRET_ID);
    form.append("response", token);
    form.append("remoteip", remoteIp);

    const result = await fetch("https://challenges.cloudflare.com/turnstile/v0/siteverify", {
      method: "POST",
      body: form,
    });

    const data = await result.json();

    if (!data.success) {
      throw new HttpError({ statusCode: 401, message: "Invalid cloudflare token" });
    }

    return data;
  } catch (error) {
    throw new HttpError({ statusCode: 500, message: "Internal Server Error" });
  }
```
can't we put this code in try catch block

**Comment by sean-brydon** [comment]
> The lack of alignment with the other items, the harsh borders, the oversized check icon, and the fact it's the only item with colour (when the button is disabled); is making this stick out a lot when it's the only thing on the page we don't need them to engage with in some way since it's automatic right?

We can set the mode on cloudflares side to be invisible -> 
CleanShot 2024-03-12 at 14 47 01

But i'd honestly recommend Managed mode as thats what they recommend themselves for the most part. 

We dont have too much control over what this looks like from what i can tell. I can try my best to style this to match our theme

**Comment by sean-brydon** [comment]
> > The lack of alignment with the other items, the harsh borders, the oversized check icon, and the fact it's the only item with colour (when the button is disabled); is making this stick out a lot when it's the only thing on the page we don't need them to engage with in some way since it's automatic right?
> 
> We can set the mode on cloudflares side to be invisible -> CleanShot 2024-03-12 at 14 47 01
> 
> But i'd honestly recommend Managed mode as thats what they recommend themselves for the most part.
> 
> We dont have too much control over what this looks like from what i can tell. I can try my best to style this to match our theme

Note: Everything is loaded in a iframe and we have no way of styling this

File: apps/web/pages/signup.tsx [hunk 9]
```diff
@@ -247,319 +260,336 @@ export default function Signup({
...
+                <PasswordField
+                  data-testid="signup-passwordfield"
+                  label={t("password")}
+                  {...register("password")}
+                  hintErrors={["caplow", "min", "num"]}
+                />
+                {/* Cloudflare Turnstile Captcha */}
+                {CLOUDFLARE_SITE_ID ? (
+                  <TurnstileCaptcha
+                    appearance="interaction-only"
```
Comment (sean-brydon) [inline apps/web/pages/signup.tsx:346]: If appearance is set to interaction-only, the widget will become only visible in cases where an interaction is required
 
 https://developers.cloudflare.com/turnstile/get-started/client-side-rendering/#appearance-modes

File: apps/web/lib/signup/getServerSideProps.tsx [hunk 10]
```diff
@@ -24,8 +24,9 @@ const querySchema = z.object({
 
 export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
   const prisma = await import("@calcom/prisma").then((mod) => mod.default);
+  await ssrInit(ctx);
```
Comment (CarinaWolli) [inline apps/web/lib/signup/getServerSideProps.tsx:27]: can this be removed too?

File: apps/web/pages/signup.tsx (hunk 9 above)
Comment (CarinaWolli) [reply apps/web/pages/signup.tsx:346]: Great find, that looks like what we want

File: apps/web/lib/signup/getServerSideProps.tsx (hunk 10 above)
Comment (sean-brydon) [reply apps/web/lib/signup/getServerSideProps.tsx:27]: No they seems to be the only way translations load on signup 🤔

**Review by CarinaWolli — COMMENTED** [review]
Looks good to me, but let's have somebody from @calcom/foundation to approve here

**Review by joeauyeung — CHANGES_REQUESTED** [review]
This is much needed. I tested with a different combination of test keys and have some  questions before I feel comfortable approving.

- When failing we don't show the Cloudflare turnstile component but the create account button is blocked. Is this what we want? It renders though when I force an interaction
Example with a failing sitekey
image
Forced interaction
image

**Review by joeauyeung — DISMISSED** [review]
Talking to @sean-brydon about my last review and I feel like my question has been answered. Since @CarinaWolli has also reviewed this PR as well I feel comfortable approving this PR with two reviews.

**Review by zomars — DISMISSED** [review]
LGTM

**Comment by zomars** [comment]
Let's fix conflicts and ensure tests are passing before merging 🙏🏽

**Review by keithwillcode — APPROVED** [review]
All checks now passing cc @zomars

**Pull request merged**